### PR TITLE
fix: split prod/dev requirements to shrink Heroku slug under 1GB

### LINKED ISSUE #975: P1: Unblock Heroku prod deploy — split requirements + aggressive slug strip
## Summary

Prod deploy to `opencreorg` is blocked: Heroku rejects the slug at **3.1G** (max **1000M**). Root cause is `sentence-transformers` pulling CUDA PyTorch into the Heroku build via `requirements.txt`.

Latest failure: https://github.com/OWASP/OpenCRE/actions/runs/29030616421

```
Compiled slug size: 3.1G is too large (max is 1000M).
Push rejected to opencreorg.
```

## Goal

Unblock Heroku prod deploy with a **tiny** prod slug: no torch/CUDA, and no importer/LLM/test tooling on the dyno. Prod serves cache-only GA + REST + auth + worker queue.

## Product decisions (confirmed for this ticket)

1. **Chat/completion on opencreorg** — OK if it 500s without `litellm` (compute/chat tooling is local/dev).
2. **OSCAL export / Google Sheets import on prod** — CLI/local only; not required on the dyno.
3. Worker jobs that call embeddings/import on Heroku may fail — matches “compute locally, prod is cache-only”.

### Files

| Action | Path |
|--------|------|
| Rewrite | `requirements.txt` (prod, minimal) |
| Create | `requirements-dev.txt` (`-r requirements.txt` + everything else) |
| Update | `Makefile`, `Dockerfile`, `Dockerfile-dev`, `.github/workflows/linter.yml`, `scripts/run-local.sh`, `scripts/update-cwe.sh` |
| Fix boot imports | `cre.py`, `application/web/web_main.py`, `application/cmd/cre_main.py`, `application/utils/spreadsheet.py` (and any other top-level pulls of heavy deps) |

### Prod `requirements.txt` (target)

Keep only what web/worker/release need **after** import fixes:

```
# server
gunicorn
Flask
Flask-Caching
flask-compress
Flask-Cors
Flask-Migrate
Flask-SQLAlchemy
SQLAlchemy
psycopg2-binary
setuptools
click
python-dotenv

# queue / cache
redis
rq

# graph / DB helpers used by db.py
neo4j
neomodel
networkx
PyYAML==6.0.1

# auth + HTTP
requests
oauthlib
google
google-api-core
google_auth_oauthlib

# OpenAPI / schemas used by web
marshmallow
apispec
pydantic>=2.4.0,<3
dacite
semver
python-markdown-maker

# optional analytics (env-gated)
posthog

typing_extensions
```

### Dev `requirements-dev.txt`

```
-r requirements.txt

# ML / Librarian
sentence-transformers>=5.0.0,<6.0.0

# embeddings / chat / import tooling (not on Procfile)
litellm
openai
vertexai
google-cloud-aiplatform
google-cloud-trace
grpcio
grpcio_status
playwright
nltk
numpy
scipy
scikit-learn
beautifulsoup4
pypdf
alive-progress
compliance-trestle
gspread
GitPython
PyGithub
xmltodict
openpyxl
Pillow
# … remaining freeze noise + lint/test (black, mypy, pytest*, coverage, etc.)
```

### Import fixes (required for aggressive strip)

Without these, dynos fail on missing modules at boot:

1. **`cre.py`** — lazy-import `coverage` (keep module load free of test-only deps)
2. **`web_main.py`** — stop top-level `import cre_main` / `alive_progress` / OSCAL / spreadsheet-gspread; lazy where used
3. **`cre_main.py`** — stop top-level `prompt_client` import; lazy inside register/embeddings/librarian
4. **`spreadsheet.py`** — don’t import `gspread` at module top if CSV export is the only prod path

**Why:** today `gunicorn cre:app` → `web_main` → `cre_main` → `prompt_client` eagerly loads `nltk` / `playwright` / `sklearn`. Chat’s “lazy import” comment is wrong relative to current boot.

### Steps

1. Break eager import chains (above)
2. Write slim prod + fat dev requirement files
3. Point local/CI/Docker-dev at `requirements-dev.txt`; Heroku stays on `requirements.txt` (buildpack default)
4. Run lint/mypy/tests
5. Redeploy — expect slug ≪ 1 GB

## Success criteria

- [ ] `requirements.txt` has no `sentence-transformers`, `torch`, `playwright`, `nltk`, `scikit-learn`, `litellm`, Google AI SDKs, test/lint tools
- [ ] `make lint` / `make mypy` / `make test` pass with `requirements-dev.txt`
- [ ] App imports cleanly with only prod deps (smoke: `python -c "from application import create_app; create_app()"` in a prod-only venv)
- [ ] Heroku deploy slug &lt; 1000M (`Deploy to OPENCREORG` green)

## Risks

- Chat / OSCAL / Sheets paths on **prod** will 500 until those deps are present — accepted for this ticket
- Over-stripping a transitive dep → add back one package
- Worker embeddings/import on Heroku will fail — accepted (local compute)

## Out of scope

- CPU-only torch pin
- Changing Librarian feature behavior beyond import laziness
- Removing neo4j from prod (still imported by `db.py`)
- Expanding `.slugignore` as the primary fix (nice-to-have only)

## Context / prior work

- Chat plan approved 2026-07-10; action first thing next day
- Related: Module C Librarian added `sentence-transformers` to the shared requirements file
- Pattern: Heroku uses `requirements.txt` by default; no deploy.yml change needed if prod file stays named that way

## Verification commands

```bash
make lint
make mypy
make test
# after merge / deploy:
# confirm Actions Deploy to OPENCREORG succeeds and slug < 1000M
```

### PR DESCRIPTION
**Pull request by northdpole**
## Summary
- Split dependencies: slim `requirements.txt` for Heroku/prod (no `sentence-transformers`/torch), fat `requirements-dev.txt` for local/CI/importers/Librarian ML.
- Lazy-import heavy modules (`prompt_client`, playwright/nltk, gspread, alive_progress, OSCAL, coverage) so `gunicorn cre:app` boots without pulling importer/ML tooling.
- Point Makefile, linter CI, and local scripts at `requirements-dev.txt`; Heroku/Dockerfile keep default `requirements.txt`.

Closes #975.

## Out of scope
- pgvector / chat similarity cutover (#977)
- Unrelated local WIP (`.yarn*`, `import_telemetry.json`)

## Test plan
- [x] `make lint`
- [x] Targeted: `python -m unittest application.tests.web_main_test application.tests.spreadsheet_test`
- [ ] CI green on this PR
- [ ] After merge: Deploy to OPENCREORG — confirm slug &lt; 1000M
- [ ] Smoke prod: `/rest/v1/standards`, `/docs`, chat (still has litellm/sklearn on prod in this PR)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by northdpole** [comment]
Review request for the GSoC Module A–C authors (GitHub only allows formal review requests for collaborators; @Pa04rth is already requested as a reviewer):

- Module A (harvester): @ParthAggarwal16
- Module B (noise / relevance filter): @manshusainishab
- Module C (Librarian): @PRAteek-singHWY

Please take a look when you can — especially anything that affects local/CI installs via `requirements-dev.txt` vs the slim Heroku `requirements.txt`.

**Comment by PRAteek-singHWY** [comment]
Hi @northdpole took a look from the Module C (Librarian) side.
Once it lands, local and CI will pull requirements-dev.txt through make install-python, and that carries 
sentence-transformers plus the numpy / scipy / scikit-learn it inherits from the base requirements.txt, so the retriever, cross-encoder, and calibration should all install fine and the librarian tests should get everything they need. 

Dropping sentence-transformers / torch from the slim prod requirements.txt is fine for us too, since the Librarian isn't actually run by the web server yet (it's opt-in CLI / dry-run) and the cross-encoder import is lazy, so prod won't try to load them at boot.

I'll re-verify locally once it's merged. 
From the C side though, seems fine.

**Comment by ParthAggarwal16** [comment]
hi @northdpole, took a look from the Module A (Harvester) side.

From what I can see, nothing here changes the harvester pipeline itself, and the dependency split doesn’t affect the parser/retriever/normalizer flow. Since the harvester only relies on the runtime dependencies that remain in the slim requirements.txt, moving the importer/ML tooling into requirements-dev.txt shouldn’t impact Module A. The lazy imports also don’t appear to touch any of the diff pipeline code paths.

I also took a quick look through the lazy import changes. Overall they look good and I don’t see any obvious missing imports or circular import issues. The new tests around the optional dependency handling are a nice addition as well.

Only one minor observation :

* The RuntimeError check based on "litellm" in the exception message feels a little brittle. A dedicated exception type for optional dependency failures might be better choice (happy to be proven wrong on this) .

Otherwise, from the Module A side this LGTM. I’ll re verify after merge as well, but I don’t see anything that should affect the harvester.

**Comment by manshusainishab** [comment]
Hi @northdpole, took a look from the Module B (noise / relevance filter) side.

The dependency split looks good. Module B's only real runtime dependency beyond the core is litellm —> the Stage 2 classifier calls Gemini through it  and since that stays in the slim requirements.txt for chat, B classifies fine on both prod and dev. We don't touch any of the ML tooling moving to requirements-dev.txt (no sentence-transformers / torch / sklearn on our side the classifier is LLM-only), so trimming those from the slim set is fine for Module B. Our tests are fully mocked (no live LLM or network), so they'll get everything they need from requirements-dev.txt in CI.

On the lazy import direction  Module B already follows it. Our llm_classifier imports litellm lazily inside the classifier constructor, and at module level it only pulls the lightweight prompt_client.llm_error_utils (the rate-limit detector), not the heavy prompt_client module. I double-checked: importing B's Stage 2 module loads neither litellm nor prompt_client (nor any ML deps) — so B won't drag importer/ML tooling into the boot path. When we wire B into the pipeline via its CLI entry point, it'll stay on that slim path.
